pkg/istio/virtualservice.go

```diff
@@ -0,0 +1,145 @@
+		vs, err := getVirtualServiceMapped(ctx.Namespace, resource.Name)
+		//vs, err := getVirtualService2(ctx, ctx.Namespace, resource.Name)
+		if err != nil {
+			if errors.IsNotFound(err) { // Not found, nothing to clean
+				break
+			}
+			ref.AddResourceStatus(model.ResourceStatus{Kind: VirtualServiceKind, Name: resource.Name, Action: model.ActionFailed})
+			break
+		}
+
+		ctx.Log.Info("Found VirtualService", "name", resource.Name)
+		mutatedVs, err := revertVirtualService(*vs)
+		if err != nil {
+			ref.AddResourceStatus(model.ResourceStatus{Kind: VirtualServiceKind, Name: resource.Name, Action: model.ActionFailed})
+			break
+		}
+		//err = setVirtualService(ctx.Namespace, &mutatedVs)
+		err = ctx.Client.Update(ctx, &mutatedVs)
+		if err != nil {
+			ref.AddResourceStatus(model.ResourceStatus{Kind: VirtualServiceKind, Name: resource.Name, Action: model.ActionFailed})
+			break
+		}
+		// ok, removed
+		ref.RemoveResourceStatus(model.ResourceStatus{Kind: VirtualServiceKind, Name: resource.Name})
+	}
+
+	return nil
+}
+
+func mutateVirtualService(vs istionetwork.VirtualService) (istionetwork.VirtualService, error) {
+
+	source := vs.Spec.Http[0]
+
+	sourceRoutes := []*v1alpha3.HTTPRouteDestination{}
+	for _, r := range source.Route {
+		sourceRoute := v1alpha3.HTTPRouteDestination{
+			Destination: &v1alpha3.Destination{
+				Host:   r.Destination.Host,
+				Port:   r.Destination.Port,
+				Subset: r.Destination.Subset + "-test",
+			},
+		}
+		sourceRoutes = append(sourceRoutes, &sourceRoute)
+	}
+
+	route := &v1alpha3.HTTPRoute{
+		Match: []*v1alpha3.HTTPMatchRequest{
+			&v1alpha3.HTTPMatchRequest{
+				Headers: map[string]*v1alpha3.StringMatch{
+					"end-user": &v1alpha3.StringMatch{MatchType: &v1alpha3.StringMatch_Exact{Exact: "jason"}},
+				},
+			},
+		},
+		Route:                 sourceRoutes,
+		Redirect:              source.Redirect,
+		AppendHeaders:         source.AppendHeaders,
+		AppendResponseHeaders: source.AppendResponseHeaders,
+		RemoveRequestHeaders:  source.RemoveRequestHeaders,
+		RemoveResponseHeaders: source.RemoveResponseHeaders,
+		CorsPolicy:            source.CorsPolicy,
+		Fault:                 source.Fault,
+		Headers:               source.Headers,
+		Mirror:                source.Mirror,
+		Retries:               source.Retries,
+		Rewrite:               source.Rewrite,
+		Timeout:               source.Timeout,
+		WebsocketUpgrade:      source.WebsocketUpgrade,
+	}
+	vs.Spec.Http = append([]*v1alpha3.HTTPRoute{route}, vs.Spec.Http...)
+
+	return vs, nil
+}
+
+func revertVirtualService(vs istionetwork.VirtualService) (istionetwork.VirtualService, error) {
+
+	for i := 0; i < len(vs.Spec.Http); i++ {
+		if strings.Contains(vs.Spec.Http[i].Route[0].Destination.Subset, "-test") {
+			vs.Spec.Http = append(vs.Spec.Http[:i], vs.Spec.Http[i+1:]...)
+			break
+		}
+	}
+	return vs, nil
+}
+
+func getVirtualService2(ctx model.SessionContext, namespace, name string) (*istionetwork.VirtualService, error) {
```

Comment:
`2`?

Reply:
OC vs Rest CLI. Hack from trying to get istio apis to do something useful.